### PR TITLE
fixed seasons

### PR DESCRIPTION
**Pull request by mediaminister**
This fixes https://github.com/pietje666/plugin.video.vrt.nu/issues/124

### CODE REVIEW

File: resources/lib/vrtplayer/vrtapihelper.py [hunk 1]
```diff
@@ -166,13 +166,17 @@ def _map_to_season_items(self, api_url, buckets, episode):
...
         for bucket in buckets:
-            season_title = bucket.get('key')
-            title = ''.join((self._kodi_wrapper.get_localized_string(32094), ' ', season_title))
+            season_key = bucket.get('key')
+            season_size = str(bucket.get('doc_count'))
+            title = ''.join((self._kodi_wrapper.get_localized_string(32094), ' ', season_key))
             season_facet = '&facets[seasonTitle]='
-            if ' ' in season_title:
-                season_title = season_title.replace(' ', '-')
+            # Workaround: seasonTitle facet doesn't support spaces and hyphens properly, luckily seasonName facet escapes this
```
Comment (dagwieers) [inline resources/lib/vrtplayer/vrtapihelper.py]: To me this is not clear. So I doubt anyone else reading this code will understand why this is needed.
Maybe it would help if we add examples in the comments for each case.

Also, why don't we *always* replace ' ' with '-', and always check for seasonKey instead ?

Comment (dagwieers) [reply resources/lib/vrtplayer/vrtapihelper.py]: So my implementation would simply do this:
```python
            season_title = bucket.get('key')
            season_size = str(bucket.get('doc_count'))
            title = ''.join((self._kodi_wrapper.get_localized_string(32094), ' ', season_title))

            # Workaround: seasonTitle facet doesn't support spaces and hyphens properly, luckily seasonName facet escapes this
            season_key = season_title.replace(' ', '-')
            season_facet = '&facets[seasonName]='
```

Comment (mediaminister) [reply resources/lib/vrtplayer/vrtapihelper.py]: We get the available season values from faceted search api with `bucket.get('key')`, and **these values always match the `seasonTitle` values**.
For instance: https://vrtnu-api.vrt.be/search?i=video&facets[programUrl]=//www.vrt.be/vrtnu/a-z/girls-talk/
We get value `1`, and `2`
Then we filter the episodes from season 1 with the following request: https://vrtnu-api.vrt.be/search?i=video&facets[programUrl]=//www.vrt.be/vrtnu/a-z/girls-talk/&facets[seasonTitle]=1

We get 5 episodes, this is correct.

And now, try to filter with `seasonName`: https://vrtnu-api.vrt.be/search?i=video&facets[programUrl]=//www.vrt.be/vrtnu/a-z/girls-talk/&facets[seasonName]=1

We get 0 episodes, because the value of `seasonName` is `2018` and doesn't match the `bucket.get('key')` value, wich is `1`.

So, `seasonName` isn't a reliable key to get the episodes from the corresponding season.

Comment (dagwieers) [reply resources/lib/vrtplayer/vrtapihelper.py]: Thanks for the detailed explanation.

I wonder if there's no better way to escape those characters, rather than assume the seasonName will related to the seasonTitle (when spaces or dashes are encountered).
I tried various possibilities to escape (or do exact matches) using seasonName, but it seems to be a starting match (so `facets[seasonTitle]=2` or `facets[seasonTitle]=201` return the same thing).

Comment (mediaminister) [reply resources/lib/vrtplayer/vrtapihelper.py]: I found a way to maximise the use of `seasonName`.

File: resources/lib/vrtplayer/vrtapihelper.py [hunk 2]
```diff
@@ -160,19 +160,27 @@ def _map_to_episode_items(self, episodes, titletype=None):
...
+        # Reverse sort seasons if program_type is 'reeksaflopend' or 'daily'
+        seasons = sorted(seasons, key=lambda k: k['key'], reverse=True) if program_type in ('reeksaflopend', 'daily') else seasons
+        for season in seasons:
+            season_key = season.get('key')
+            season_episode_count = str(season.get('doc_count'))
+            title = ''.join((self._kodi_wrapper.get_localized_string(32094), ' ', season_key))
+            # API workaround: season_key matches seasonTitle facet, but seasonTitle behaves as a starting match regex
+            # and therefore sometimes mixes up episodes from different seasons, instead we prefer to use seasonName facet
+            # which behaves as an exact match regex and in most cases matches season_key when spaces are replaced by dashes.
+            if len(episode.get('seasonTitle')) > 1 and episode.get('seasonTitle').replace(' ', '-') == episode.get('seasonName'):
```
Comment (dagwieers) [inline resources/lib/vrtplayer/vrtapihelper.py]: Not sure if the `len > 1` is a good identifier. Can we not do i.e. `int()` instead ?

File: resources/lib/vrtplayer/vrtapihelper.py (hunk 1 above)
Comment (dagwieers) [reply resources/lib/vrtplayer/vrtapihelper.py]: While testing various possible syntaxis I got this error:
```
TypeError: facet_value.indexOf is not a function
   at module.exports (/var/app/current/lib/queryBuilders/vrt-video/search.js:83:37)
   at Object.makeESQuery (/var/app/current/lib/queryBuilders/index.js:6:20)
   at handler (/var/app/current/lib/requestHandlers/search.js:21:42)
   at Layer.handle [as handle_request] (/var/app/current/node_modules/express/lib/router/layer.js:95:5)
   at trim_prefix (/var/app/current/node_modules/express/lib/router/index.js:312:13)
   at /var/app/current/node_modules/express/lib/router/index.js:280:7
   at Function.process_params (/var/app/current/node_modules/express/lib/router/index.js:330:12)
   at next (/var/app/current/node_modules/express/lib/router/index.js:271:10)
   at cors (/var/app/current/node_modules/cors/lib/index.js:179:7)
   at /var/app/current/node_modules/cors/lib/index.js:229:17
```
trying: https://vrtnu-api.vrt.be/search?i=video&facets[programUrl]=//www.vrt.be/vrtnu/a-z/universiteit-van-vlaanderen/&facets=seasonTitle:201
This gives a little insight into what was used for the API frontend, i.e. nodejs and queryBuilders

**Comment by dagwieers** [comment]
I like this new implementation, using more than just the "bucket".

File: resources/lib/vrtplayer/vrtapihelper.py (hunk 2 above)
Comment (mediaminister) [reply resources/lib/vrtplayer/vrtapihelper.py]: Not sure what you mean. `seasonTitle` is a string.

Comment (dagwieers) [reply resources/lib/vrtplayer/vrtapihelper.py]: I get that, but I assumed that you do `len(episode.get('seasonTitle')) > 1` to filter out integer values as seasonTitle. So I wonder if turning it into an integer wouldn't be a better check.

Comment (mediaminister) [reply resources/lib/vrtplayer/vrtapihelper.py]: No, not necessary, I found out that just checking the season data of one episode is not failsafe when VRT uses a different `seasonName` naming convention for the other seasons.
This was the case with "Down the road" https://vrtnu-api.vrt.be/search?i=video&size=150&facets[programName]=down-the-road
So, I tried to filter out as much values that are not affected by the "starting match regex problem" and luckily this fixes the "Down the road" problem.

Comment (dagwieers) [reply resources/lib/vrtplayer/vrtapihelper.py]: I don't understand what this has to do with a length of more than 1 though. So I have a big problem if we create conditional statements without a clue to what it relates to. It makes it impossible to refactor in the future not knowing if you break something that worked before.

I assumed you wanted to filter out when seasonTitle was 1-9, now I don't know why you do this. A season could be 10 or 11 too.

Comment (dagwieers) [reply resources/lib/vrtplayer/vrtapihelper.py]: From the subtitle you can see the VRT is using the seasonTitle and not the seasonName, so I guess we need to do the same. Somehow I think the solution is in the displayOptions.

PS I discovered it's not actually a "starting match regex", for `2018-2019` the following match: `2018`, `201`, `2`, but also `2019`. What doesn't work is `19` or `18`. A lot of special characters seems to terminate the string (including `-`, `+`, `*` or `|`) although if any special character is repeated before the [A-z0-9] string, these characters are ignored. A very weird implementation...

Comment (mediaminister) [reply resources/lib/vrtplayer/vrtapihelper.py]: Values with a length of one are immune for the "starting match regex problem". I filter them out so they are handled by the `seasonTitle` facet.

I don't have a better solution right now to fix "Down the road" season listing:  https://vrtnu-api.vrt.be/search?i=video&size=150&facets[programName]=down-the-road
Down the road has two seasons with `seasonTitle` values: `1` and `2` and `seasonName` values: `2018` and `2`
From `season.get('key')` we get  `1` and `2`  as input values.

So, the most logical thing is to use `seasonTitle` facet here.

Comment (dagwieers) [reply resources/lib/vrtplayer/vrtapihelper.py]: Well, that is why I suggested to check if they are integers instead, or maybe even check if the value is `[A-z0-9]+`.

Comment (mediaminister) [reply resources/lib/vrtplayer/vrtapihelper.py]: I don't understand, for instance value `1` and value `13` are affected by the "starting match regex problem"

Comment (dagwieers) [reply resources/lib/vrtplayer/vrtapihelper.py]: Ok, so my concern here is that if we *have to* implement workarounds for very specific corner-cases, we need to clearly identify these corner-cases in comments (with actual examples) or we have no chance in the future to refactor/improve this code ever. (Ideally we would have proper unit-tests for this at some point).

Now that I understand a bit better why this is more complex, let me suggest some changes in this PR.

PS I'd rather avoid these complexities, but it's clear we have no clue how to escape special characters (which seem to terminate the string matching) at this time, if this is at all possible :-/

Comment (mediaminister) [reply resources/lib/vrtplayer/vrtapihelper.py]: Feel free to fix this better. I don't have time this week to look at this again.

Comment (dagwieers) [reply resources/lib/vrtplayer/vrtapihelper.py]: The "better" fix will mostly be documenting the corner-cases, but I need to test them myself first, and that won't happen this week...

Comment (mediaminister) [reply resources/lib/vrtplayer/vrtapihelper.py]: It seems seasonTitle facet is implemented in the Elasticsearch backend as a partial match. This is not fixable in a good way using api requests. I fixed this again in a different way.

**Comment by dagwieers** [comment]
I made a few cosmetic changes to limit the indentation or improve readability.
Thanks for these fixes/improvements !

PS @pietje666 I think we need a v1.7.2 release with all the stuff we improved recently.
